tools/wptrunner/wptrunner/executors/executorservo.py

```diff
@@ -283,3 +286,95 @@ class ServoDriverProtocol(WebDriverProtocol):
 
 class ServoWdspecExecutor(WdspecExecutor):
     protocol_cls = ServoDriverProtocol
+
+
+class ServoTimedRunner(TimedRunner):
+    def run_func(self):
+        try:
+            self.result = True, self.func(self.protocol, self.url, self.timeout)
+        except Exception as e:
+            message = getattr(e, "message", "")
+            if message:
+                message += "\n"
+            message += traceback.format_exc(e)
+            self.result = False, ("INTERNAL-ERROR", message)
+        finally:
+            self.result_flag.set()
+
+    def set_timeout(self):
+        pass
+
+
+class ServoCrashtestExecutor(ProcessTestExecutor):
+    convert_result = crashtest_result_converter
+
+    def __init__(self, browser, server_config, binary=None, timeout_multiplier=1,
+                 screenshot_cache=None, debug_info=None, pause_after_test=False,
+                 **kwargs):
+        ProcessTestExecutor.__init__(self,
+                                     browser,
+                                     server_config,
+                                     timeout_multiplier=timeout_multiplier,
+                                     debug_info=debug_info)
+
+        self.pause_after_test = pause_after_test
+        self.protocol = ConnectionlessProtocol(self, browser)
+        self.tempdir = tempfile.mkdtemp()
+        self.hosts_path = write_hosts_file(server_config)
+
+    def do_test(self, test):
+        timeout = (test.timeout * self.timeout_multiplier if self.debug_info is None
+                   else None)
+
+        test_url = self.test_url(test)
+        self.prefs =  test.environment.get('prefs', {})
+
+        success, data = ServoTimedRunner(self.logger, self.do_crashtest, self.protocol,
+                                         test_url, timeout, self.extra_timeout).run()
+
+        if success:
+            return self.convert_result(test, data)
+
+        return (test.result_cls(*data), [])
+
+    def do_crashtest(self, protocol, url, timeout):
+        env = os.environ.copy()
+        env["HOST_FILE"] = self.hosts_path
+        env["RUST_BACKTRACE"] = "1"
+
+        args = [
+            "--hard-fail", "-u", "Servo/wptrunner",
+            "-x",
+            "-Z", "replace-surrogates", "-z", url,
+        ]
+        for stylesheet in self.browser.user_stylesheets:
```

Comment:
This looks like some code we should extract into a helper rather than duplicate.